Refactor documentation generation code. Extend fallback rules to include group.

cmnbroad opened the pull request:
Changes based on review of gatk usage.

lbergelson commented on src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java:
```diff
@@ -307,35 +306,41 @@ private DocumentedFeatureObject getFeatureForClass(final Class<?> clazz) {
-                    groupSummary = "No group summary available";
+                }
+            }
+            if (summary.length() == 0) {
```
I always prefer `isEmpty` to `length == 0` checks.

cmnbroad replied:
Done.

lbergelson commented on src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java:
```diff
@@ -307,35 +306,41 @@ private DocumentedFeatureObject getFeatureForClass(final Class<?> clazz) {
+                    summary = clProps.oneLineSummary();
+                }
+                if (summary.length() == 0) {
+                    logger.warn("No summary declared for: " + clazz.getCanonicalName() + ". Trying javadoc fallback.");
```
Do we no longer need the backup text of "No X available" or to worry about them being null?

cmnbroad replied:
They're never null because the annotation attributes now all have default values of "". We no longer want to populate the strings with "No X declared" here since further on, when we have access to the javadoc, we'll use the empty string to determine if we should attempt to fall back to get a summary from that. I'll add a comment explaining that here. Ultimately this all needs refactoring.

lbergelson replied:
Is it fair to say that we don't care about people who perversely over write the default with null?

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DefaultDocumentedFeatureHandler.java:
```diff
@@ -137,7 +133,7 @@ protected void addHighLevelBindings(
             final Map<String, Object> root)
     {
-        root.put("name", currentWorkUnit.classDoc.name());
+        root.put("name", currentWorkUnit.clazz.getSimpleName());
```
Yes.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DefaultDocumentedFeatureHandler.java:
```diff
@@ -256,11 +252,15 @@ protected void addCommandLineArgumentBindings(
-        final StringBuilder summaryBuilder = new StringBuilder();
-        for (final Tag tag : currentWorkUnit.classDoc.firstSentenceTags()) {
-            summaryBuilder.append(tag.text());
+        // If no summary was found from annotations, use the javadoc if there is any
```
Agreed, though I think that the refactoring this code so badly needs should be done in a separate PR since it needs bigger modeling changes (i.e., DocumentedFeatureObject should probably go away completely, and DocWorkUnit should just contain the truth).

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocumentedFeatureObject.java:
```diff
@@ -54,11 +54,14 @@ public DocumentedFeatureObject(
-    public String summary() { return summary; }
+
+    public String getSummary() { return summary; }
+    public void putSummary(final String newSummary) { summary = newSummary; }
```
Done.

cmnbroad commented:
Added two more commits. The first is for the code review comments. The second is refactoring:

Removed classes:
- DocumentedFeatureObject (just use DocumentedFeature annotation directly)
- DocUtils (merged with DocletUtils)

Renamed classes:
- DocumentedFeatureHandler -> DocWorkUnitHandler
- DefaultDocumentedFeatureHandler -> DefaultDocWorkUnitHandler

Allow consumers to override creation of HelpDoclet, DocWorkUnit and DocWorkUnitHandler
Use builtin javadoc destination dir argument instead of custom argument

lbergelson commented on src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java:
```diff
@@ -143,17 +147,16 @@ else if (!settingsDir.isDirectory())
      */
     public static int optionLength(final String option) {
```
I don't see this method used anywhere, here or in the gatk pull request.  Is it something we still need?  Is it part of the doclet contract?

cmnbroad replied:
Yes, its called by javadoc.

lbergelson commented on src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java:
```diff
@@ -225,125 +291,44 @@ private void processDocs(final RootDoc rootDoc) {
     }
 
+    private CommandLineProgramProperties getCommandLineProgramPropertiesForClass(final Class<?> clazz) {
+        return clazz.isAnnotationPresent(CommandLineProgramProperties.class) ?
```
this is redundant i think, `getAnnotation()` returns null if it isn't present

cmnbroad replied:
Yes, removed.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java:
```diff
@@ -27,84 +27,189 @@
+     */
+    protected String getDescription(final DocWorkUnit currentWorkUnit) {
+        return Arrays.stream(currentWorkUnit.getClassDoc().inlineTags())
+                .filter(t -> getTagPrefix() == null ? true : !t.name().startsWith(getTagPrefix()))
```
Makes more sense. Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java:
```diff
@@ -27,84 +27,189 @@
+
+    /**
+     * Return the description to be used for the work unit. We need to manually strip
+     * out any inline custom javdoc tags since we don't those in the summary.
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java:
```diff
@@ -27,84 +27,189 @@
+     * out any inline custom javdoc tags since we don't those in the summary.
+     *
+     * @param currentWorkUnit
+     * @return Descrition to be used or the work unit.
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java:
```diff
@@ -285,38 +347,39 @@ private String getTagPrefix() {
 
     // Add top level bindings for the default instances for each plugin descriptor
-    // NOTE: The default freemarker template provided by Barlcay has no references to plugins, since they're
+    // NOTE: The default Freemarker template provided by Barlcay has no references to plugins, since they're
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java:
```diff
@@ -26,106 +26,186 @@
     /**
-     * The documentedFeatureObject that lead to this Class being in Doc
+     * Get the JavDoc ClassDoc for this work unit.
+     * @return ClassDoc for this work unit. Will noe be null.
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java:
```diff
@@ -26,106 +26,186 @@
+    public Class<?> getClazz() { return clazz; }
+
+    /**
+     * Populate the property map for this wokr unit by delegating to the documented feature handler for this work unit.
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java:
```diff
@@ -26,106 +26,186 @@
+    };
+
+    /**
+     * Get the template to be used for this work unit. Delegates to th feature handler.
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java:
```diff
@@ -26,106 +26,186 @@
+                return commandLineProperties.programGroup().newInstance();
+            } catch (IllegalAccessException | InstantiationException e) {
+                logger.warn(
+                        String.format("Cant instantiate program group class to retrieve summary for group %s for class %s",
```
Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocWorkUnitHandler.java:
```diff
@@ -0,0 +1,113 @@
+ * used for each documented feature, and populates the template property map for that template.
+ */
+public abstract class DocWorkUnitHandler {
+    private HelpDoclet doclet;
```
Yes. Done.

cmnbroad commented on src/main/java/org/broadinstitute/barclay/help/DocWorkUnitHandler.java:
```diff
@@ -0,0 +1,113 @@
+/*
```
Good point. I'm removing from all of the doc files in this PR.
